platforms/software/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java

```diff
@@ -90,44 +96,60 @@ public enum PathBase {
-    private DistributionType distributionType = WrapperDefaults.DISTRIBUTION_TYPE;
-    private String archivePath = WrapperDefaults.ARCHIVE_PATH;
-    private PathBase archiveBase = WrapperDefaults.ARCHIVE_BASE;
     private final Property<Integer> networkTimeout = getProject().getObjects().property(Integer.class);
-    private boolean distributionUrlConfigured = false;
     private final boolean isOffline = getProject().getGradle().getStartParameter().isOffline();
+    private final RegularFileProperty batchScript;
+    private final RegularFileProperty propertiesFile;
 
     public Wrapper() {
         getValidateDistributionUrl().convention(WrapperDefaults.VALIDATE_DISTRIBUTION_URL);
+        getScriptFile().convention(getProject().getObjects().fileProperty()).fileValue(getFileOperations().file(WrapperDefaults.SCRIPT_PATH));
+        getJarFile().convention(getProject().getObjects().fileProperty()).fileValue(getFileOperations().file(WrapperDefaults.JAR_FILE_PATH));
+        getDistributionPath().convention(DEFAULT_DISTRIBUTION_PARENT_NAME);
+        getDistributionBase().convention(WrapperDefaults.DISTRIBUTION_BASE);
+        getDistributionType().convention(WrapperDefaults.DISTRIBUTION_TYPE);
+        getArchivePath().convention(WrapperDefaults.ARCHIVE_PATH);
+        getArchiveBase().convention(WrapperDefaults.ARCHIVE_BASE);
+        getGradleVersion().convention(GradleVersion.current().getVersion());
+        this.batchScript = getObjectFactory().fileProperty().fileProvider(getScriptFile().getLocationOnly().map(file -> WrapperGenerator.getBatchScript(file.getAsFile())));
+        this.propertiesFile = getObjectFactory().fileProperty().fileProvider(getJarFile().getLocationOnly().map(file -> WrapperGenerator.getPropertiesFile(file.getAsFile())));
     }
 
     @TaskAction
     void generate() {
-        File jarFileDestination = getJarFile();
-        File unixScript = getScriptFile();
+        File jarFileDestination = getJarFile().get().getAsFile();
+        File unixScript = getScriptFile().getAsFile().get();
         FileResolver resolver = getFileLookup().getFileResolver(unixScript.getParentFile());
         String jarFileRelativePath = resolver.resolveAsRelativePath(jarFileDestination);
-        File propertiesFile = getPropertiesFile();
+        File propertiesFile = getPropertiesFile().get().getAsFile();
         Properties existingProperties = propertiesFile.exists() ? GUtil.loadProperties(propertiesFile) : null;
+        DistributionType distributionType = getDistributionType().get();
+        gradleVersionResolver.setGradleVersionString(getGradleVersion().get());
 
         checkProperties(existingProperties);
-        validateDistributionUrl(propertiesFile.getParentFile());
+
+        String distributionUrl = getDistributionUrl().isPresent()
+            ? getDistributionUrl().get()
+            : WrapperGenerator.getDistributionUrl(gradleVersionResolver.getGradleVersion(), distributionType);
```

Comment:
❓ Could we make this a convention on `distributionUrl`?

Also, could you make sure that the scenario described in the following issue becomes an option with the changes?
That would be a nice way to validate the benefits of having `Property` types here.

See:
* #27295

Reply:
Yes, using convention makes sense. There is some complication due to `gradleVersionResolver` keeping version in it's state, but I workaround that for now.

Using is now possible
```
distributionUrl = gradleVersion.zip(distributionType) { version, type ->
    "(https://example.com/distributions/gradle-${version}-${type}.zip"
}
```
is possible with providers: we now also use that also in the convention, see:
https://github.com/gradle/gradle/blob/d9c29b7862d8b5a02f9f77dd91b0563117c86b54/platforms/software/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java#L115C1-L116C1

Reply:
Tested also locally. You can use:
```
distributionUrl = gradleVersion.zip(distributionType) { version, type ->
    "https://example.com/distributions/gradle-${version}-${type.name.lowercase()}.zip"
}
```